# This patch makes sure to write "VBE2" at the start of the vesa_general_info structure,
# which makes the kernel detect some video modes it would otherwise not detect.
#
# On my netbook (HP Mini 110-3500), without this patch, the biggest video mode that is detected
# is 800x600, while the "proper" resolution should be 1024x600. The latter is only detected
# if "VBE2" is written in the structure, as this patch does.
#
# This shouldn't break support for other computers, and I'm pretty sure it's well-tested,
# as GRUB2 has code[1] for this, and it's mentioned in Ralf Brown's Interrupt List[2].
#
# Links:
# [1]: http://git.savannah.gnu.org/cgit/grub.git/tree/grub-core/video/i386/pc/vbe.c#n522
# [2]: http://www.delorie.com/djgpp/doc/rbinter/it/77/0.html

diff --git a/arch/x86/boot/vesa.h b/arch/x86/boot/vesa.h
index 468e444..0379059 100644
--- a/arch/x86/boot/vesa.h
+++ b/arch/x86/boot/vesa.h
@@ -30,6 +30,7 @@ struct vesa_general_info {
 } __attribute__ ((packed));
 
 #define VESA_MAGIC ('V' + ('E' << 8) + ('S' << 16) + ('A' << 24))
+#define VBE2_MAGIC ('V' + ('B' << 8) + ('E' << 16) + ('2' << 24))
 
 struct vesa_mode_info {
 	u16 mode_attr;		/* 0 */
diff --git a/arch/x86/boot/video-vesa.c b/arch/x86/boot/video-vesa.c
index ba3e100..f464546 100644
--- a/arch/x86/boot/video-vesa.c
+++ b/arch/x86/boot/video-vesa.c
@@ -40,6 +40,8 @@ static int vesa_probe(void)
 
 	video_vesa.modes = GET_HEAP(struct mode_info, 0);
 
+	vginfo.signature = VBE2_MAGIC;
+
 	initregs(&ireg);
 	ireg.ax = 0x4f00;
 	ireg.di = (size_t)&vginfo;
